kame/kame/mcastsend/mcastsend.1
.\" Copyright (C) 1995, 1996, 1997, 1998, and 1999 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd Jul 29, 1999
.Dt MCASTSEND 1
.Os KAME
.\"
.Sh NAME
.Nm mcastsend
.Nd send multicast packets
.\"
.Sh SYNOPSIS
.Nm mcastsend
.Op Fl l Ar hop_limit
.Op Fl i Ar interface
.Op Fl s Ar scope
.Op Fl h
.Ar address
.Ar port
.Op Ar file
.\"
.Sh DESCRIPTION
The 
.Nm
utility reads the given
.Ar file
.Po
or the standard input if no file is specified
.Pc
and sending them to the multicast
.Ar address
you specified.
Also you must specify UDP
.Ar port
number or service name
.Pq Xr services 5 .
.Pp
The options are as follows:
.Bl -tag -width Fl
.It Fl l Ar hop_limit
Specify the number of hop limit.
By default, it depends on the system wide value,
net.inet6.ip6.defmcasthlim.
.It Fl i Ar interface
Specify the interface name to use sending packets.
By default, kernel will decide.
.It Fl s Ar scope
Specify the hint in order to decision of source IP address to send packets.
By default, global scope is applied.
If you don't specify
.Ar interface ,
.Nm
ignores this option.
.Ar scope
takes one of the following numeric:

.Bl -tag -width "14 " -compact
.It 1	node local
.It 2	link local
.It 5	site local
.It 14	global
.El
.It Fl h
Show this help message.
.El
.\"
.Sh RETURN VALUES
.Nm
will exit with positive value on errors.
.\"
.Sh SEE ALSO
.Xr mcastread 1 ,
.Xr sysctl 8
.\"
.Sh HISTORY
The
.Nm
command first appeared in WIDE/KAME IPv6 protocol stack kit.
.\"
.Sh BUGS
IPv4 is not supported.


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/mcastsend/mcastsend.c
/*
 * Copyright (C) 1995, 1996, 1997, 1998, and 1999 WIDE Project.
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
#include <sys/param.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/sockio.h>
#include <sys/ioctl.h>
#include <net/if.h>
#include <netinet/in.h>
#include <netdb.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <err.h>
#include <errno.h>
#include <ifaddrs.h>

#ifndef IPV6_ADDR_SCOPE_RESERVED
#define IPV6_ADDR_SCOPE_RESERVED	0x00
#endif
#ifndef IPV6_ADDR_SCOPE_NODELOCAL
#define IPV6_ADDR_SCOPE_NODELOCAL	0x01
#endif
#ifndef IPV6_ADDR_SCOPE_LINKLOCAL
#define IPV6_ADDR_SCOPE_LINKLOCAL	0x02
#endif
#ifndef IPV6_ADDR_SCOPE_SITELOCAL
#define IPV6_ADDR_SCOPE_SITELOCAL	0x05
#endif
#ifndef IPV6_ADDR_SCOPE_ORGLOCAL
#define IPV6_ADDR_SCOPE_ORGLOCAL	0x08	/* just used in this file */
#endif
#ifndef IPV6_ADDR_SCOPE_GLOBAL
#define IPV6_ADDR_SCOPE_GLOBAL		0x0e
#endif

char *pname;
char *addr, *port;
char *file = NULL;
char *ifname = NULL;
int hlim = 0;
int scope_limit = IPV6_ADDR_SCOPE_RESERVED;
u_int debug = 0;

void Usage __P((void));
int parse __P((int, char **));
int bind_srcaddr __P((struct addrinfo *, int, char *, int));

void
Usage()
{
	printf("Usage: %s [-l hlim] [-i ifname [-s scope]] "
		"addr port [file]\n", pname);
	printf("\tscope: 1, 2, 5 or 14\n");
}

int
main(ac, av)
	int ac;
	char **av;
{
	int s;
	struct addrinfo hints, *res;
	int error;

	parse(ac, av);

	memset(&hints, 0, sizeof(hints));
	hints.ai_family = PF_UNSPEC;
	hints.ai_socktype = SOCK_DGRAM;
	error = getaddrinfo(addr, port, &hints, &res);
	if (error != 0)
		errx(1, "%s", gai_strerror(error));

	switch (res->ai_family) {
	case AF_INET6:
	case AF_INET:
		break;
	default:
		errno = EAFNOSUPPORT;
		err(1, "%s %s", addr, port);
	}

	s = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
	if (s < 0)
		err(1, "socket");

	/* set interface to send and bind. */
	if (ifname != NULL) {
		int ifindex;
		struct in_addr a;

		ifindex = if_nametoindex(ifname);
		if (ifindex == 0)
			err(1, "if_nametoindex");
		switch (res->ai_family) {
		case AF_INET6:
			error = setsockopt(s, IPPROTO_IPV6, IPV6_MULTICAST_IF,
					&ifindex, sizeof(ifindex));
			if (error < 0)
				err(1, "setsockopt(IPV6_MULTICAST_IF)");
			break;
		case AF_INET:
			/* XXX uses "0.0.0.0/8 for interface index" API */
			a.s_addr = htonl(ifindex);
			error = setsockopt(s, IPPROTO_IP, IP_MULTICAST_IF,
					&a, sizeof(a));
			if (error < 0)
				err(1, "setsockopt(IP_MULTICAST_IF)");
			break;
		}

		error = bind_srcaddr(res, s, ifname, scope_limit);
		if (error < 0)
			errx(1, "Failed to bind source address.");
	}

	/* set hop-limit */
	if (hlim != 0) {
		u_char ttl;

		switch (res->ai_family) {
		case AF_INET6:
			error = setsockopt(s, IPPROTO_IPV6, IPV6_MULTICAST_HOPS,
					&hlim, sizeof(hlim));
			if (error < 0)
				err(1, "setsockopt(IPV6_MULTICAST_HOPS)");
			break;
		case AF_INET:
			ttl = hlim & 0xff;
			error = setsockopt(s, IPPROTO_IP, IP_MULTICAST_TTL,
					&ttl, sizeof(ttl));
			if (error < 0)
				err(1, "setsockopt(IP_MULTICAST_TTL)");
			break;
		}
	}

	/* main */
    {
	int fd = STDIN_FILENO;	/* default: stdin */
	int len;
	char buf[BUFSIZ];

	if (file != NULL) {
		fd = open(file, O_RDONLY);
		if (fd < 0)
			err(1, "open");
	}

	while ((len = read(fd, buf, sizeof(buf))) > 0) {
		len = sendto(s, buf, len, 0, res->ai_addr, res->ai_addrlen);
		if (len < 0) {
			close(fd);
			close(s);
			err(1, "send");
		}
	}

	close(fd);
    }

	close(s);
	exit(0);
}

int
bind_srcaddr(res, so, ifname, scope_limit)
	struct addrinfo *res;
	int so;
	char *ifname;
	int scope_limit;
{
	struct ifaddrs *ifa, *ifap, *match;
	struct in6_addr *a;
	int scope_max;

	/* XXX no scope support for IPv4 yet */
	if (res->ai_family != AF_INET6)
		return 0;

	if (getifaddrs(&ifap) != 0)
		return -1;

	match = NULL;
	scope_max = IPV6_ADDR_SCOPE_NODELOCAL;

	for (ifa = ifap; ifa; ifa = ifa->ifa_next) {
		if (strcmp(ifa->ifa_name, ifname) != 0)
			continue;
		if (res->ai_family != ifa->ifa_addr->sa_family)
			continue;

		a = &((struct sockaddr_in6 *)ifa->ifa_addr)->sin6_addr;

		if (IN6_IS_ADDR_LOOPBACK(a))
			continue;

		if (IN6_IS_ADDR_LINKLOCAL(a) &&
		    scope_limit >= IPV6_ADDR_SCOPE_LINKLOCAL &&
		    scope_max < IPV6_ADDR_SCOPE_LINKLOCAL) {
			scope_max = IPV6_ADDR_SCOPE_LINKLOCAL;
			match = ifa;
			continue;
		}

		if (IN6_IS_ADDR_SITELOCAL(a) &&
		    scope_limit >= IPV6_ADDR_SCOPE_SITELOCAL &&
		    scope_max < IPV6_ADDR_SCOPE_SITELOCAL) {
			scope_max = IPV6_ADDR_SCOPE_SITELOCAL;
			match = ifa;
			continue;
		}

		if (scope_max < IPV6_ADDR_SCOPE_GLOBAL &&
		    scope_limit >= IPV6_ADDR_SCOPE_GLOBAL) {
			scope_max = IPV6_ADDR_SCOPE_GLOBAL;
			match = ifa;
			continue;
		}
	}

	if (!match) {
		freeifaddrs(ifap);
		return -1;
	}

	if (bind(so, match->ifa_addr, match->ifa_addr->sa_len) < 0) {
		freeifaddrs(ifap);
		return -1;
	}

	freeifaddrs(ifap);
	return 0;
}

int
parse(ac, av)
	int ac;
	char **av;
{
	int c;

	pname = *av;

	while ((c = getopt(ac, av, "hdl:i:s:")) != -1) {
		switch (c) {
		case 'l':
			hlim = atoi(optarg);
			break;
		case 'i':
			ifname = optarg;
			break;
		case 's':
			scope_limit = atoi(optarg);
			break;
		case 'd':
			debug++;
			break;
		case 'h':
			Usage();
			exit (0);
		default:
			Usage();
			exit (1);
		}
	}

	/* check ifname and scope */
	if (ifname == NULL && scope_limit != IPV6_ADDR_SCOPE_RESERVED) {
		warnx("Ignore scope.");
	} else
	if (scope_limit != IPV6_ADDR_SCOPE_RESERVED
	 && scope_limit != IPV6_ADDR_SCOPE_NODELOCAL
	 && scope_limit != IPV6_ADDR_SCOPE_LINKLOCAL
	 && scope_limit != IPV6_ADDR_SCOPE_SITELOCAL
	 && scope_limit != IPV6_ADDR_SCOPE_GLOBAL) {
		warnx("Invalid scope.\n");
		Usage();
		exit (1);
	}

	if (ifname != NULL && scope_limit == IPV6_ADDR_SCOPE_RESERVED)
		scope_limit = IPV6_ADDR_SCOPE_GLOBAL;

	ac -= optind;
	av += optind;

	switch (ac) {
	case 3:
		file = *(av + 2);
	case 2:
		addr = *av;
		port = *(av + 1);
		break;
	default:
		Usage();
		exit (1);
	}

	return 0;
}
